### PR TITLE
Add Yggdrasil support

### PR DESCRIPTION
**Pull request by prusnak**
Heavily inspired by https://github.com/bitcoin/bitcoin/pull/23077 + https://github.com/bitcoin/bitcoin/pull/23324 + https://github.com/bitcoin/bitcoin/pull/23175 which add CJDNS support

# Yggdrasil overview 

Yggdrasil is an implementation of a new name-independent routing scheme and functions as an end-to-end encrypted IPv6 network. It shares very similar goals to CJDNS. It uses an address range of `0200::/8` (as opposed to `fc00::/8` used by CJDNS).

# Motivation

Despite Yggdrasil is being considered an alpha quality software, it seems to be much more maintained than CJDNS.

CJDNS contributions chart:

Screenshot 2021-11-17 at 10 02 06

Yggdrasil contributions chart:

Screenshot 2021-11-17 at 10 02 21

It is also much easier to get it Yggdrasil working on variety of systems and architectures (macOS, Windows, Android, ARM64, MIPS) because it is written in pure Go, while it is hard to get CJDNS working on anything else than x86_64 Linux.

# Yggdrasil resources

* homepage: https://yggdrasil-network.github.io
* more about Yggdrasil history: https://yggdrasil-network.github.io/2019/01/09/history.html
* FAQ: https://yggdrasil-network.github.io/faq.html
* source code: https://github.com/yggdrasil-network/yggdrasil-go
* releases: https://github.com/yggdrasil-network/yggdrasil-go/releases

### CODE REVIEW

**Comment by prusnak** [comment]
@vasild @jonatack Would you like to experiment with Yggdrasil too?

**Comment by sipa** [comment]
I set up an Yggdrasil router as well: `[200:b9b6:8c66:f88d:7043:e30b:d803:c3e2]:8333` (it's not running this branch, but I don't think that matters).

**Review by ghost — CHANGES_REQUESTED** [review]
UPDATE: No opinions on this PR: https://github.com/bitcoin/bitcoin/pull/23531#issuecomment-974300160

> Despite Yggdrasil is being considered an alpha quality software, it seems to be much more maintained than CJDNS

~~NACK~~

Reasons: 

1. Alpha project should not be used for mainnet. For experiments maybe implementing only for testnet makes sense.
2. ipv4, ipv6, Tor, i2p and cjdns already exist. Adding more networks will add complexity based on combinations and possibility of bugs. 
3. Few other reasons mentioned here: https://github.com/bitcoin/bitcoin/pull/23077#issuecomment-926626650
4. GNUnet will be a better alternative if/when one more network makes sense

**Comment by laanwj** [comment]
Mind that this is a P2P BIP-level change, I think you'll want to write a (short) ancillary BIP to BIP155 to propose this new assigned number.

File: test/functional/feature_proxy.py
Comment (jonatack) [inline test/functional/feature_proxy.py:112]: 0a33d6c only had a quick look, so to be confirmed, but it looks like you need to do something like this for the functional test to pass
```diff
 class ProxyTest(BitcoinTestFramework):
     def set_test_params(self):
-        self.num_nodes = 5
+        self.num_nodes = 6
         self.setup_clean_chain = True
 
     def setup_nodes(self):
@@ -109,7 +109,9 @@ class ProxyTest(BitcoinTestFramework):
             ['-listen', f'-proxy={self.conf2.addr[0]}:{self.conf2.addr[1]}','-proxyrandomize=1'],
             [],
             ['-listen', f'-proxy={self.conf1.addr[0]}:{self.conf1.addr[1]}','-proxyrandomize=1',
-                '-cjdnsreachable', '-yggdrasilreachable']
+                '-cjdnsreachable'],
+            ['-listen', f'-proxy={self.conf1.addr[0]}:{self.conf1.addr[1]}','-proxyrandomize=1',
+                '-yggdrasilreachable'],
         ]
```

**Review by jonatack — COMMENTED** [review]
Haven't looked into or tested yggdrasil yet. Concept ACK on additional networks available to users if that strengthens the robustness of the peer-to-peer network. Debug build is clean for me. Some initial light feedback.

File: src/bitcoin-cli.cpp
Comment (jonatack) [reply src/bitcoin-cli.cpp:52]: (maybe just "ygg"?)

File: test/functional/feature_proxy.py
Comment (prusnak) [reply test/functional/feature_proxy.py:112]: Addressed in 578ffd92b774f33391d77b9c00366a30b27d9611

**Comment by vasild** [comment]
The addresses used by Yggdrasil are considered routable by Bitcoin Core. That is - just like any other normal IPv6 address, publicly routable on the global internet. They will be gossiped. This is in contrast to CJDNS which uses unroutable addresses from RFC 4193 and thus need special handling in Bitcoin Core.

I have to think more and review carefully this PR, but it may be that it is not necessary and Yggdrasil "just works" with current `master`.

> It uses an address range of `0200::/8`

According to https://yggdrasil-network.github.io/faq.html#will-yggdrasil-conflict-with-my-network-routing it uses `0200::/7` (starts with `02` or `03`).

> it is hard to get CJDNS working on anything else than x86_64 Linux

Hmm, CJDNS works for me on FreeBSD/x86_64 and on my OpenWRT router - Linux/mips (there is also an yggdrasil package for OpenWRT).

**Comment by prusnak** [comment]
> I have to think more and review carefully this PR, but it may be that it is not necessary and Yggdrasil "just works" with current `master`.

That would great news actually!

**Comment by sipa** [comment]
If we add support for this, I do think we want to at least include logic to treat Yggdrasil's addresses separately for the purpose of group mapping (e.g. like Tor, treat all of Yggdrasil as 1 or a few network groups, rather than using the normal IPv6 logic). We could do that with or without treating it as a separate network:
* With a separate network, as this PR currently does, it's a somewhat cleaner separation, we get somewhat better failures (non-Yggdrasil-connected nodes would not even bother trying to connect, rather than trying and failing), and if we ever get a network extension to negotiate supported networks, also somewhat better gossip (by preferential relay among compatible nodes). The downside is that it requires a BIP or BIP addendum to reserve a BIP155 network id.
* Without a separate network, it's a smaller change, and we'd implement the group mapping directly on the 0200::/8 range in IPv6, essentially hardcoding the knowledge that that range belongs to Yggdrasil.

I have a slight preference for the first approach.

**Comment by vasild** [comment]
> ... treat Yggdrasil's addresses separately for the purpose of group mapping ...

Right!

> I have a slight preference for the first approach.

Me too.

**Comment by jb55** [comment]
Interesting... Concept ACK

**Comment by jamesob** [comment]
Concept ACK, will review

**Comment by prusnak** [comment]
> I think you'll want to write a (short) ancillary BIP to BIP155 to propose this new assigned number.

PR to bips repo in https://github.com/bitcoin/bips/pull/1238

**Comment by vasild** [comment]
Concept ACK

Without this PR, `master` is able to connect to yggdrasil addresses and will gossip them (good). With minimal changes (this PR), we can get (thanks, @sipa):

* Proper reporting, like in `getpeerinfo` RPC (to print `yggdrasil` instead of `ipv6`)
* Proper network grouping - don't do nonsensical netmasks on those addresses
* Don't try to connect to such addresses if `-someconfigoption` is not set to indicate yggdrasil connectivity is available
* Selective gossip

**Comment by sipa** [comment]
@prayank23 Moving the discussion from https://github.com/bitcoin/bips/pull/1238#issuecomment-974190303 here, because I think it's more a question about Bitcoin Core's implemention.

I'm not convinced we should add this either, because of the low maturity of the project. But I think the discussion is a bit more nuanced than just looking at that.

The name of this PR is a bit misleading, for starters. This isn't adding support for yggdrasil to Bitcoin Core in the sense of having the project depend on it, or add direct integration (as was the case for tor and i2p, where significant amount of special code were needed to let Bitcoin Core interact with the specific routing software for those networks). What it does add is a small amount of code to make Bitcoin Core's network stack *aware* of the fact that a particular subrange of the IPv6 space corresponds to Yggdrasil, and should be treated slightly differently. Interacting with the Yggdrasil network is done independently, by the user installing an Yggdrasil router, which integrates in the OS, and transparently routes yggdrasil-compliant IP addresses over the Yggdrasil network - and this happens (and can be used) whether or not Bitcoin Core has specific support for it.

The advantages are:
* Better behavior in case *anyone* starts rumouring such addresses on the network, regardless of whether the node's operator considers them using Yggdrasil to be a good idea or not.
* Better behavior for anyone actually using Yggdrasil. I personally think that more distinct ways to connect nodes in general is a good thing, as it contributes to partition resistance of the network.

The disadvantages are:
* Extra code to maintain, obviously.
* Connectivity issues in case there are problems/upgrades with yggdrasil, for those using it.

I think both of the disadvantages are actually tiny. The amount of code being added here is similar to what is needed for cjdns; just making the network stack aware of the existence of this separate range. Connectivity issues could be problem in case someone is exclusively connecting through yggdrasil, but I don't think anyone is arguing for that (or should; just like you shouldn't exclusively use automatic tor or i2p connections). If yggdrasil goes down (either due to a network-wide failure, or incompatible upgrades, or individual nodes crashing)... bitcoin core will make new connections; interrupted connections are nothing special. Given that Bitcoin Core is in no way directly talking to the yggdrasil router, I'm not worried about it causing bugs/crashes in Bitcoin Core itself.

**Comment by prayank23** [comment]
@sipa That makes sense. Thanks for explaining everything in detail.

I retract my NACK and have no opinion on this pull request. Will follow the progress, research more about yggdrasil and do tests though.

File: src/bitcoin-cli.cpp
Comment (prusnak) [reply src/bitcoin-cli.cpp:52]: Column size set to 10 in 011d0af8082d5d94b0619fedf36db4e016f55791

**Comment by ShadowJonathan** [comment]
Hey everyone, **I would not recommend adding Yggdrasil support**, not on the basis of principle, but on the basis of the nature of the project itself, let me explain.

Yggdrasil is largely an experimental academic protocol/software, aimed to research routing at scale, the intention (of its author @Arceliar) is to - eventually - wrap it up, place a v1 on it, and write a paper, but there is sofar no intention to keep up support after that.

In the mean time, there is also no guarantees of minor version incompatibility, so this could be quite unstable to use.

Instead, i would like to direct everyone's attention to a spin-off project, [pinecone](https://github.com/matrix-org/pinecone/), made with the same current routing methods, but intended to be supported for a long time. Pinecone also doesn't work via ipv6 addresses, but full-on peer hashes, and is intended to be embeddable.

However, as Bitcoin is a large project on the internet with many eyes on it, i would heavily recommend getting into dialogue with Pinecone's developers before proceeding with adding support.

TL;DR: Yggdrasil is experimental and academic, while a project derivative of it, Pinecone, is aiming to be actively developed in the future.

**Comment by prusnak** [comment]
> wrap it up, place a v1 on it, and write a paper, but there is sofar no intention to keep up support after that.

Can you please comment on that? @neilalexander @arceliar

**Comment by ShadowJonathan** [comment]
@prusnak please also read my disclaimer at the bottom, it's the intention i got from the authors, my point being that yggdrasil is more an academic project than a continuous software one. @neilalexander works on yggdrasil on the latter basis, but has split off efforts into Pinecone, primarily inspired by Yggdrasil.

**Comment by jsarenik** [comment]
Looks good to me. Skimmed through the patch, everything seems trivial there. Running yggdrasil (often shortened to `ygg` which I would prefer in the column names instead of widening them) since early 2017.

```
$ echo $COLUMNS
80
```

**Comment by Arceliar** [comment]
> > wrap it up, place a v1 on it, and write a paper, but there is sofar no intention to keep up support after that.
> 
> Can you please comment on that? @neilalexander @Arceliar

Just to clarify, v1 only means that I'm satisfied the R&D work is "done" and things are ready to be documented and published. That doesn't mean the project will be abandoned, just that it's no longer a platform for ongoing research (or at least, not on compact routing).

**Comment by ShadowJonathan** [comment]
Alright, i've somewhat skimmed the changes made here, and I do think they're pretty resistant to yggdrasil's version changes, but i'd like to re-iterate that major protocol changes *can* happen between versions, and that addresses may not always be guaranteed to be addressable from the local node, until v1

**Comment by neilalexander** [comment]
> Can you please comment on that? @neilalexander @Arceliar

Given that all the PR seems to do is to add the understanding that `0200::/7` is unofficially Yggdrasil address space for the purposes of peer connectivity and configurability, I don’t really think there’s an issue here. Even if we reach a supposed v1, it also doesn’t mean that the current network will go away overnight. 

Yggdrasil protocol iterations also don’t necessarily mean address space changes either — we changed address space once, from `fd00::/8` to `0200::/7`, back in June 2018 to avoid conflicting with other projects and private address space. We have never needed to change it since, even though multiple protocol breaks have happened in that time, and have no plans to change it for the foreseeable either.

As long as the only dependency here is on the address space and not any other aspect of the protocol then I would think this is safe enough to do.

**Comment by ShadowJonathan** [comment]
> As long as the only dependency here is on the address space and not any other aspect of the protocol then I would think this is safe enough to do.

Agreed, admittedly my concern mainly was if and when yggdrasil would be imported as a dependency into existing binaries, but that doesn't seem to be the case here.

**Review by jsarenik — COMMENTED** [review]
Just some unimportant nit ideas. Looks good to me as it is.

File: src/bitcoin-cli.cpp
Comment (prusnak) [reply src/bitcoin-cli.cpp]: "yggdrasil" renamed to "ygg" in da2803cf1c20d5b63a6be6ecc3f703ea84b59649

Comment (prusnak) [reply src/bitcoin-cli.cpp]: "yggdrasil" renamed to "ygg" in da2803cf1c20d5b63a6be6ecc3f703ea84b59649

File: src/bitcoin-cli.cpp
Comment (prusnak) [reply src/bitcoin-cli.cpp]: "yggdrasil" renamed to "ygg" in da2803cf1c20d5b63a6be6ecc3f703ea84b59649

File: src/net.cpp
Comment (prusnak) [reply src/net.cpp]: done in 0b3dd4140e34ca2a87b813bb810958d70b5160d8

**Review by jsarenik — APPROVED** [review]
Thank you @prusnak for the changes!

**Comment by jsarenik** [comment]
Tested ACK a85cd9187d33bd9a2859aabe28c656825da0e539

```
$ ./src/bitcoind -version
Bitcoin Core version v22.99.0-a85cd9187d33
$ ./src/bitcoind -help | grep -B1 ygg
       Make outgoing connections only through network <net> (ipv4, ipv6, onion,
       i2p, cjdns, ygg). Incoming connections are not affected by this
--

  -yggdrasilreachable
```

`make check` passed. Will soon start it with a few more patches on top to prove it works in the wild.

**Comment by jsarenik** [comment]
```
$ bitcoin-cli -signet -netinfo
Bitcoin Core client v22.99.0-878d8e1c0661 signet - server 70016/Satoshi:22.99.0/

         ipv4   onion     ygg   total   block
in          0      25       2      27
out         6       4       0      10       2
total       6      29       2      37

Local addresses
78.98.6.214                                                        port  38333    score      4
s7fcvn5rblem7tiquhhr7acjdhu7wsawcph7ck44uxyd6sismumemcyd.onion     port  38333    score      4
200:5ecb:6b8a:d837:30e:8a01:c858:612b                              port  38333    score      1
300:5ecb:6b8a:d837::1                                              port  38333    score      1
300:5ecb:6b8a:d837::2                                              port  38333    score      1
300:5ecb:6b8a:d837::3                                              port  38333    score      3
```

I have the same settings for mainnet and testnet, and inbound connections get in, but I do not see the addresses in `-netinfo` like on signet. Will check another day.

 * https://ex.signet.bublina.eu.org/node-details
 * https://ln.anyone.eu.org/signet.txt

File: src/netaddress.cpp
Comment (jsarenik) [inline src/netaddress.cpp:439]: shauld be `an Yggdrasil address`

File: src/netaddress.cpp
Comment (jsarenik) [inline src/netaddress.cpp:483]: or `0x03`? It is written in the comment, but not in the condition.

**Review by jsarenik — CHANGES_REQUESTED** [review]
I will have a look again. So far I found this.

**Comment by jsarenik** [comment]
After using time-machine (`v0.20.2`) I think all my current confusion comes from the `bitcoin-cli -netinfo` which did not exist back then. But there is definitely a `bitcoin-cli -netinfo` issue here, though it is not related to this PR at all.

**Comment by jsarenik** [comment]
UPDATE: Everything works well for me now. I had to set `discover=1` because I use `externalip=<current_IPv4>` and it disables `discover` by default. Interestingly the behavior differs with 'signet' network. But that's not related now. Full ACK for this PR.

File: src/netaddress.cpp
Comment (jsarenik) [reply src/netaddress.cpp:483]: Probably a false-positive. Everything works well. I was just looking for something, some detail which would fix it for me, but found it to be `discover=1` in the end.

Comment (prusnak) [reply src/netaddress.cpp:483]: Right. `((m_addr[0] & 0xfe) != 0x02))` is the same as `((m_addr[0] != 0x02) && (m_addr[0] != 0x03))

File: src/netaddress.cpp
Comment (prusnak) [reply src/netaddress.cpp]: Done in aa5c9779bdd1c816899c0704ee1d965a57d6f036

**Comment by jsarenik** [comment]
Tested ACK 234731ae7

Running on https://ln.anyone.eu.org (before last rebase I was rebasing it myself to `master` branch with [one little unrelated patch on top](https://github.com/jsarenik/bitcoin/commit/280e2a1dbb393521dd9f763673837dac96d86a28)) for more than a month now. Works as expected. Thank you!

File: src/qt/locale/bitcoin_en.xlf
Comment (laanwj) [inline src/qt/locale/bitcoin_en.xlf:3064]: Please don't update the translation files in a PR, we leave this until specific times to minimize merge collisions.

**Comment by carnhofdaki** [comment]
@prusnak please rebase to current `master`

**Comment by majestrate** [comment]
i've had a full node reachable via yggdrasil for about a year (`[201:4c36:45a0:4b59:d0d0:9d6a:311d:fca7]:8333`) and I didn't know there was a PR for adding Yggdrasil awareness so i'll chuck in my 2 cents on this topic. There are quite a few overlay networks and the number of them seems likely to only increase in the future. Perhaps consolidating all of these kinds into a distinct network type for overlay networks would be a more future-proof approach. e.g. a network type just for overlay networks advertising a mapping with key of "insert-arbitrary-overlay-network-name-here" to ip-ranges-used/gtld-used, so peers can pick and choose which intersection they'd want to utilize. I think this would reduce the number of PRs for "hey add support for my special network layer please" everyone in bitcoin core would need to review and filter, which would likely be a win for everyone involved (especially bitcoin core reviewers).

**Comment by prusnak** [comment]
I agree with the above comment. 

Also I am closing this PR, but anyone feel free to take it as a base to for another PR which adds overlay networks in more generic way.

**Comment by prusnak** [comment]
> Network ids are cheap (there are 256),

While networks ids are cheap, getting them allocated is quite hard: see https://github.com/bitcoin/bips/pull/1238

**Comment by majestrate** [comment]
> While networks ids are cheap, getting them allocated is quite hard: see bitcoin/bips#1238

that is indeed the issue at hand. hence why suggesting a new network id as a catchall for "overlay network" with a peer address containing a mapping of "overlay-name" to "reachable-range/reachable-gtld"

**Comment by jsarenik** [comment]
I use this patch on be.anyone.eu.org node and the reason is, there is also an IPv6 tunnel on the machine but I do not want `bitcoind` to use it and still would like it to use the Yggdrasil address, which works correctly now. My branch rebased to `v24.0rc2` is at https://github.com/jsarenik/bitcoin/tree/jsn/yggdrasil-24 (and includes one extra last last commit which fixes locales on Alpine Linux).

**Comment by jonatack** [comment]
> > it is hard to get CJDNS working on anything else than x86_64 Linux
> 
> Hmm, CJDNS works for me on FreeBSD/x86_64 and on my OpenWRT router - Linux/mips (there is also an yggdrasil package for OpenWRT).

I also found CJDNS to be easy to install on macOS (M1 Ventura in my testing) with `brew install cjdns`.

**Comment by prusnak** [comment]
> I also found CJDNS to be easy to install on macOS (M1 Ventura in my testing) with `brew install cjdns`.

cjdns has horrible build system - you need Python 3.9 (not 3.10 or newer), Rust and NodeJS just to compile something which is mostly a C99 code 🤮 It's notoriously hard to get it working on new platforms.

**Pull request closed without merging**
